sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpMessageConverter.cs

```diff
@@ -752,31 +766,72 @@ private static ArraySegment<byte> ReadStreamToArraySegment(Stream stream)
-            stream.CopyTo(memStream, StreamBufferSizeInBytes);
+            switch (stream)
+            {
+                case BufferListStream bufferListStream:
+                    return bufferListStream.ReadBytes((int)stream.Length);
 
-            return new ArraySegment<byte>(memStream.ToArray());
+                default:
+                {
+                    using var memStream = new MemoryStream(StreamBufferSizeInBytes);
+                    stream.CopyTo(memStream, StreamBufferSizeInBytes);
+                    return new ArraySegment<byte>(memStream.ToArray());
+                }
+            }
         }
 
         /// <summary>
-        ///   Converts a stream to a set of memory bytes.
+        ///   Reads the data body of an AMQP message, transforming it for use
+        ///   as the body of an <see cref="EventData" /> instance.
         /// </summary>
         ///
-        /// <param name="stream">The stream to read and capture in memory.</param>
+        /// <param name="body">The body data set of an AMQP message.</param>
         ///
-        /// <returns>The set of memory bytes containing the stream data.</returns>
+        /// <returns>A <see cref="BinaryData" /> representation of the <paramref name="body"/>.</returns>
         ///
-        private static ReadOnlyMemory<byte> ReadStreamToMemory(Stream stream)
+        private static BinaryData ReadAmqpDataBody(IEnumerable<Data> body)
         {
-            if (stream == null)
+            var writer = new ArrayBufferWriter<byte>();
+
+            foreach (var data in body)
             {
-                return ReadOnlyMemory<byte>.Empty;
+                var dataBytes = GetDataBytes(data);
+                dataBytes.CopyTo(writer.GetMemory(dataBytes.Length));
+
+                writer.Advance(dataBytes.Length);
             }
 
-            using var memStream = new MemoryStream(StreamBufferSizeInBytes);
-            stream.CopyTo(memStream, StreamBufferSizeInBytes);
+            return (writer.WrittenCount > 0)
+                ? BinaryData.FromBytes(writer.WrittenMemory)
+                : new BinaryData(Array.Empty<byte>());
+        }
 
-            return new ReadOnlyMemory<byte>(memStream.ToArray());
+        /// <summary>
+        ///   Gets the bytes that comprise an AMQP data instance.
+        /// </summary>
+        ///
+        /// <param name="data">The data to read the bytes from.</param>
+        ///
+        /// <returns>The set of bytes extracted from the <paramref name="data" />.</returns>
+        ///
+        private static byte[] GetDataBytes(Data data)
+        {
+            switch (data.Value)
+            {
+                case byte[] byteArray:
+                    return byteArray;
+
+                case ArraySegment<byte> segment:
+                {
+                    var byteArray = new byte[segment.Count];
+                    Array.ConstrainedCopy(segment.Array, segment.Offset, byteArray, 0, segment.Count);
```

Comment:
Are we sure this is correct? Aren't we effectively now copying the data twice for no added value since ArrayBufferWriter.CopyTo already copies the data?

https://github.com/Azure/azure-sdk-for-net/pull/19980

Reply:
My assumption is that this was being done for a reason, as I recall there being discussion around that change in Service Bus as well, but the copy is still being made.   I based that block on the current [`GetByteArray`](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageExtensions.cs#L51) implementation in Service Bus.   

@JoshLove-msft:  Do you have any thoughts here?

Reply:
I think the code flow is different between SB and EH. In SB, we don't use ArrayBufferWriter when converted into SBMessage (perhaps we should normalize here), so there isn't this duplicate copy.

Reply:
Makes sense. No, it was just that we were not doing the separate copy with ArrayBufferWriter, and I missed that subtlety in the review. Thanks @danielmarbach for catching it!